documentation/RELEASE_NOTES.md

```diff
@@ -18,6 +18,33 @@ should also be read to understand what has changed since earlier releases.
 <!-- Insert new items immediately below here ... -->
 
 
+### Timeouts for Unit Test Programs
+
+Most unit test programs that are run by the EPICS build system using
+`make runtests` or `make tapfiles` are now only allowed to run for a limited
+time period. If a test program runs for more than the timeout period (by
+default 5 minutes, 300 seconds) without exiting it will be killed after that
+period has elapsed. The limit can be extended (or shortened) for individual
+tests as necessary by adding a line to the Makefile which runs them, setting
+the environment variable `EPICS_UNITTEST_TIMEOUT` to a value in seconds while
+building the test (`.t`) script, like this:
+
+```
+  TESTSCRIPTS_HOST += hourLongTest.t
+  hourLongTest.t: export EPICS_UNITTEST_TIMEOUT=3600
+```
```

Comment:
I don't think I like encouraging hard coded overrides like this.

The author of a test can in principle determine a minimum execution time.  However, timeout is about a maximum execution time, which may vary widely depending on the runtime environment.  So I would expect that `$EPICS_UNITTEST_TIMEOUT` will always be set from CI configuration.  Practically speaking, based on the longest observed test runtime.

I suppose timeout configuration could be split, with test authors able to provide a minimum time, with maybe a multiplier provided from CI environment.  However, I'm not sure this complexity would be worthwhile.  Current best practice encourages short unittests anyway.

Reply:
Having some kind of CI-specific multiplier wouldn't really work when the test runs for a specific elapsed time, e.g. epicsMessageQueueTest, and I agree that trying to do some kind of combination adds unnecessary complexity.

A test author has to be able to override the default timeout for an individual test. The way this branch is currently coded the environment variable is checked by `makeTestfile.pl` when it is generating the `.t` script to set the default timeout written into the script, but the script itself _also_ checks for it when it's called by `make runtests` or `make tapfiles` and if provided will use that value instead of the compiled-in timeout even if that's shorter. Maybe it should display a warning if it sees that, or always use the longer of the two values?

Remember that an author's Makefile can always override any value provided by the CI environment anyway. However if they use the documented approach that you're commenting on of only setting the value when building that specific target their setting will only override the CI when the `.t` script is built, not when it's run, so the CI's value would still be used at runtime. The author can always override the CI if they want to though, you can't prevent a Makefile from doing this, which will take effect for all targets including `make tapfiles`:

```makefile
  export EPICS_UNITTEST_TIMEOUT=3600
```

Reply:
I see that I've been getting confused by the wording.  As I read it, what you're trying to say is:

Unittests based on `.t` Makefile rules (eg. `TESTSCRIPTS_HOST += $(TESTS:%=%.t)`) will now have a default timeout of 5 minutes.  A test process will be killed if it runs longer.  This __default__ timeout can be overridden for individual tests like: ...

Reply:
I rewrote the notes, hopefully this is easier to understand:

### Timeouts for Unit Test Programs

The unit test programs that are run by the `make runtests` or `make tapfiles`
commands get executed by a `.t` wrapper script which is normally generated by
the EPICS `makeTestfile.pl` program. Those generated wrapper scripts now
impose a time-limit on the test program they execute, and will kill it if it
runs for longer than 1000 seconds (16 minutes 40) without exiting. That
time-limit can be changed for any such test by modifying the Makefile which
creates and runs the `.t` wrapper script.

Setting the environment variable `EPICS_UNITTEST_TIMEOUT` to the desired
number of seconds while the Makefile is generating the test script changes the
timeout in that script. For example:

```
  TESTSCRIPTS_HOST += hourLongTest.t
  hourLongTest.t: export EPICS_UNITTEST_TIMEOUT=3600
```

When selecting such a timeout remember that different Continuous Integration
systems such as GitHub Actions and Appveyor run on processors with different
speeds, so allow enough head-room for slower systems to complete the test.

Test programs written directly in Perl as a `.plt` script should implement a
similar timeout for themselves. The "netget" test in Base does this in a way
that works on Windows as well as Unix-like hosts.